[Workflow] Add support for executing custom workflow definition validators during the container compilation

lyrixx opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix #54034
| License       | MIT


---

Usage:

```yaml
framework:
    workflows:
        article:
            definition_validators:
                - App\Workflow\Validator\ArticleValidator
```

```php
class ArticleValidator implements DefinitionValidatorInterface
{
    public function validate(Definition $definition, string $name): void
    {
        if (!$definition->getMetadataStore()->getMetadata('title')) {
            throw new InvalidDefinitionException(sprintf('The workflow metadata title is missing in Workflow "%s".', $name));
        }
    }
}
```

This code will ensure the title is defined here:
```yaml
framework:
    workflows:
        article:
            metadata:
                # title: Manage article # Commented, so it will throw an exception
```

![image](https://github.com/user-attachments/assets/949c5c69-a677-4e8d-be10-f974e1574455)

lyrixx committed:
[Workflow] Add support for executing custom workflow definition validators during the container compilation

nicolas-grekas commented:
Do you have real world examples of validation rules? I suppose you enable those validators only in the "dev" env?
I don't like the implementation, DI extensions are not meant for such things. A compiler pass would be a better fit, but if one day we need validators to support proper DI themselves, we'll have to make them services. And then, no need for any explicit configuration anymore since autoconfiguration could do the job.
Then remains the "when" do we run this. What about calling during cache warmup?

lyrixx commented:
> Do you have real world examples of validation rules? 

Here is the code my client wrote

<details><summary>WorkflowValidator.php</summary>
<p>

```php
<?php

namespace App\Validator;

use App\Service\WorkflowService;
use Symfony\Component\Workflow\Definition;
use Symfony\Component\Workflow\Exception\InvalidDefinitionException;
use Symfony\Component\Workflow\Validator\DefinitionValidatorInterface;

class WorkflowValidator implements DefinitionValidatorInterface
{
    private const VALID_PLACE_TYPES = [
        'Information',
        'UserChoice',
        'Print',
        'Modification',
        'ChooseRepair',
        'DoneRepair',
        'ReasonAndPhoto',
        'Photo',
        'ProductInformation',
        'Conditional',
        'Final',
    ];

    public function validate(Definition $definition, string $name): void
    {
        $metadataStore = $definition->getMetadataStore();
        foreach ($definition->getPlaces() as $place) {
            $transitions = $this->getForwardTransitionsByPlace($definition, $place);
            $placeMetadata = $metadataStore->getPlaceMetadata($place);

            if (! key_exists('type', $placeMetadata)) {
                throw new InvalidDefinitionException(sprintf('[%s] place "%s" has no type - %s', $name, $place, json_encode($placeMetadata)));
            }

            if (! \in_array($placeMetadata['type'], self::VALID_PLACE_TYPES, true)) {
                throw new InvalidDefinitionException(sprintf('[%s] place "%s" has a unknow type "%s"', $name, $place, $placeMetadata['type']));
            }

            switch ($placeMetadata['type']) {
                case 'Information':
                case 'Modification':
                case 'ChooseRepair':
                case 'ReasonAndPhoto':
                case 'Photo':
                case 'Print':
                    if (\count($transitions) !== 1) {
                        throw new InvalidDefinitionException(sprintf('workflow: "%s" place "%s" of type "%s" should have one outgoing transitions. found: "%s"', $name, $place, $placeMetadata['type'], $this->display($transitions)));
                    }
                    break;
                case 'UserChoice':
                    if (\count($transitions) === 0) {
                        throw new InvalidDefinitionException(sprintf('workflow: "%s" place "%s" of type "%s" should have at least one outgoing transition.', $name, $place, $placeMetadata['type']));
                    }
                    foreach ($transitions as $transition) {
                        $transitionMeta = $metadataStore->getTransitionMetadata($transition);
                        if (! \array_key_exists('text', $transitionMeta)) {
                            throw new InvalidDefinitionException(sprintf('workflow: "%s" transition "%s" should have a text', $name, $transition->getName()));
                        }
                        if (! \array_key_exists('type', $transitionMeta)) {
                            throw new InvalidDefinitionException(sprintf('workflow: "%s" transition "%s" should have a type', $name, $transition->getName()));
                        }
                    }
                    break;
                case 'Conditional':
                    if (\count($transitions) === 0) {
                        throw new InvalidDefinitionException(sprintf('workflow: "%s" place "%s" of type "%s" should have at least one outgoing transition.', $name, $place, $placeMetadata['type']));
                    }
                    break;
            }
        }
    }

    private function getForwardTransitionsByPlace(Definition $definition, string $place): array
    {
        return array_filter($definition->getTransitions(), function ($transition) use ($place) {
            return \in_array($place, $transition->getFroms(), true) && $transition->getName() !== WorkflowService::CANCEL_TRANSITION;
        });
    }

    private function display(array $transitions): string
    {
        return implode(',', array_map(function ($transition) {
            return $transition->getName();
        }, $transitions));
    }
}
```

</p>
</details> 

> I suppose you enable those validators only in the "dev" env?

I have no opinion on this one. If the definition is valid in dev, It'll be in prod. And vice versa

> DI extensions are not meant for such things

We already do some validation in the DI extension, but it could be moved to a compiler pass. I'm okay with that. More over, I used compiler pass in the project because the extension point does not exist yet.

<details><summary>Kernel.php</summary>
<p>


```php
class Kernel extends BaseKernel implements CompilerPassInterface
{
    use MicroKernelTrait;

    public function process(ContainerBuilder $container): void
    {
        // Validate all workflows
        foreach ($container->findTaggedServiceIds('workflow') as $id => $_) {
            $workflowDefinition = $container->getDefinition($id);

            $definitionArgument = $workflowDefinition->getArgument(0);
            if (! $definitionArgument instanceof Reference) {
                throw new RuntimeException('The first argument of the workflow service must be a reference.');
            }

            // warning: usually, one should not use the container to get a
            // service during the compilation phase But i'm an engineer, i know
            // what i'm doing
            $definition = $container->get((string) $definitionArgument);
            $name = $workflowDefinition->getArgument(3);

            (new WorkflowValidator())->validate($definition, $name);
        }
    }
}
```


</p>
</details> 

> And then, no need for any explicit configuration anymore since autoconfiguration could do the job.

This is not so simple. One validator could apply on one workflow but not another one. [ref](https://github.com/symfony/symfony/blob/e49f84b1207030dc9747e798a3030060a7a257a6/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L1045-L1050)

But we could pass the workflow name along the definition. But it required to update the Interface... deprecation layer... a bit boring, but why not!

>Then remains the "when" do we run this. What about calling during cache warmup?

I'm lacking knowledge in cache warmup! Is it always run, even after `rm -rf var/cache` ? Theses checks **must** be run only once, during the container building

nicolas-grekas commented:
> we could pass the workflow name along the definition

Why would we need that? We could create a WorkflowValidationCacheWarmer (and make it non-optional to answer your last question), and a compiler pass that'd configure this cach-warmer so that it knows about the names of the workflows?

lyrixx commented:
I closed the other PR. This one is the right way to do it

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1053,6 +1053,10 @@ private function registerWorkflowConfiguration(array $config, ContainerBuilder $
             $validator->validate($realDefinition, $name);
 
+            foreach ($workflow['definition_validators'] as $validatorClass) {
+                (new $validatorClass())->validate($realDefinition, $name);
```
The DefinitionValidatorInterface does not guarantee that the class has a constructor that can be called without any argument.

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -418,6 +420,23 @@ private function addWorkflowSection(ArrayNodeDefinition $rootNode): void
+                                    ->arrayNode('definition_validators')
+                                        ->beforeNormalization()
+                                            ->ifString()
+                                            ->then(fn ($v) => [$v])
```
do we need this ?

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -418,6 +420,23 @@ private function addWorkflowSection(ArrayNodeDefinition $rootNode): void
+                                        ->prototype('scalar')
+                                            ->cannotBeEmpty()
+                                            ->validate()
+                                                ->ifTrue(fn ($v) => !class_exists($v) && !interface_exists($v, false))
```
why accepting interfaces ? That won't work when instantiating them.

lyrixx commented:
@stof thanks for the review. 

* I added some code to ensure we can instantiate the validator
* I added some tests
* I removed unnecessary support for using string instead of array of strings in configuration
* I rebased to fix conflicts.

stof commented on src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md:
```diff
@@ -32,7 +34,6 @@ CHANGELOG
  * Add the `allowed_recipients` option for mailer to allow some users to receive
    emails even if `recipients` is defined.
  * Reset env vars when resetting the container
-
```
this should be reverted

lyrixx replied:
Oups, fixed

lyrixx commented:
This PR is ready 👍🏼

lyrixx commented:
ping @nicolas-grekas. 

I know you don't like it. But the PR has already +1 year. It's quite simple and solve a real use case. 

I know you would prefer to have DI enabled on the validator, but it cannot. It's all about validating **DI**. Thus the DI is not setup yet. 

More over, the validator is only about configuration. There is no need for external services.

nicolas-grekas reviewed (commented):
Could the check be moved to a compiler pass? That'd allow validating workflows that aren't defined using semantic config at least.

lyrixx commented:
> Could the check be moved to a compiler pass? That'd allow validating workflows that aren't defined using semantic config at least.

Sure. How should I transfert the data? DI Tag?

nicolas-grekas commented:
An attribute on the `workflow` tag yes?

lyrixx commented:
Ok, Now everything is done in a compiler pass. Failing tests are not related.
Ready for review

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md:
```diff
@@ -25,6 +25,8 @@ CHANGELOG
  * Deprecate `RateLimiterFactory` autowiring aliases, use `RateLimiterFactoryInterface` instead
  * Allow configuring compound rate limiters
+ * Add support for executing custom workflow definition validators during the
+   container compilation
```
```suggestion
 * Support executing custom workflow validators during container compilation
```

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -436,6 +438,23 @@ private function addWorkflowSection(ArrayNodeDefinition $rootNode): void
+                                        ->prototype('scalar')
+                                            ->cannotBeEmpty()
+                                            ->validate()
+                                                ->ifTrue(fn ($v) => !class_exists($v))
```
```suggestion
                                                ->ifTrue(static fn ($v) => !class_exists($v))
```

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -436,6 +438,23 @@ private function addWorkflowSection(ArrayNodeDefinition $rootNode): void
+                                                ->thenInvalid('The validation class %s does not exist.')
+                                            ->end()
+                                            ->validate()
+                                                ->ifTrue(fn ($v) => !is_a($v, DefinitionValidatorInterface::class, true))
```
```suggestion
                                                ->ifTrue(static fn ($v) => !is_subclass_of($v, DefinitionValidatorInterface::class))
```

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -436,6 +438,23 @@ private function addWorkflowSection(ArrayNodeDefinition $rootNode): void
+                                            ->cannotBeEmpty()
+                                            ->validate()
+                                                ->ifTrue(fn ($v) => !class_exists($v))
+                                                ->thenInvalid('The validation class %s does not exist.')
```
missing sprintf?
```suggestion
                                                ->thenInvalid('The validation class "%s" does not exist.')
```

lyrixx replied:
Nope :) It was already good

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -436,6 +438,23 @@ private function addWorkflowSection(ArrayNodeDefinition $rootNode): void
+                                            ->end()
+                                            ->validate()
+                                                ->ifTrue(fn ($v) => !is_a($v, DefinitionValidatorInterface::class, true))
+                                                ->thenInvalid(\sprintf('The validation class %%s is not an instance of %s.', DefinitionValidatorInterface::class))
```
```suggestion
                                                ->thenInvalid(\sprintf('The validation class "%%s" is not an instance of %s.', DefinitionValidatorInterface::class))
```

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -436,6 +438,23 @@ private function addWorkflowSection(ArrayNodeDefinition $rootNode): void
+                                                ->thenInvalid(\sprintf('The validation class %%s is not an instance of %s.', DefinitionValidatorInterface::class))
+                                            ->end()
+                                            ->validate()
+                                                ->ifTrue(fn ($v) => 1 <= (new \ReflectionClass($v))->getConstructor()?->getNumberOfRequiredParameters())
```
```suggestion
                                                ->ifTrue(fn ($v) => 1 <= (new \ReflectionClass($v))->getConstructor()?->getNumberOfRequiredParameters())
```
```suggestion
                                                ->ifTrue(static fn ($v) => (new \ReflectionClass($v))->getConstructor()?->getNumberOfRequiredParameters()
```

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -436,6 +438,23 @@ private function addWorkflowSection(ArrayNodeDefinition $rootNode): void
+                                            ->end()
+                                            ->validate()
+                                                ->ifTrue(fn ($v) => 1 <= (new \ReflectionClass($v))->getConstructor()?->getNumberOfRequiredParameters())
+                                                ->thenInvalid('The validation class %s constructor must not have any arguments.')
```
```suggestion
                                                ->thenInvalid('The validation class "%s" constructor must not have any arguments.')
```

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1142,14 +1144,26 @@ private function registerWorkflowConfiguration(array $config, ContainerBuilder $
             }
 
+            // Validation
+            $workflow['definition_validators'][] = match($workflow['type']) {
```
```suggestion
            $workflow['definition_validators'][] = match ($workflow['type']) {
```

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1142,14 +1144,26 @@ private function registerWorkflowConfiguration(array $config, ContainerBuilder $
+            $workflow['definition_validators'][] = match($workflow['type']) {
+                'state_machine' => Workflow\Validator\StateMachineValidator::class,
+                'workflow' => Workflow\Validator\WorkflowValidator::class,
+                default => throw new \LogicException('Invalid workflow type.'),
```
```suggestion
                default => throw new \LogicException(\sprintf('Invalid workflow type "%s".', $workflow['type'])),
```

nicolas-grekas reviewed (approved):
Just minor tweaks on my side, LGTM.

lyrixx commented:
@nicolas-grekas Thanks for the review. I addressed your comments, and I also added "resource tracking" on Validator file

nicolas-grekas commented:
The failure looks related.

lyrixx commented:
Thanks, fixed. You suggestion was wrong about `""`

fabpot commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -431,11 +433,28 @@ private function addWorkflowSection(ArrayNodeDefinition $rootNode): void
+                                            ->end()
+                                            ->validate()
+                                                ->ifTrue(static fn ($v) => 1 <= (new \ReflectionClass($v))->getConstructor()?->getNumberOfRequiredParameters())
+                                                ->thenInvalid('The validation class %s constructor must not have any arguments.')
```
```suggestion
                                                ->thenInvalid('The "%s" validation class constructor must not have any arguments.')
```

lyrixx replied:
I updated this to 
```php
->thenInvalid('The %s validation class constructor must not have any arguments.')
```

without double quote, because they are automatically added by the config component.

fabpot commented on src/Symfony/Bundle/FrameworkBundle/composer.json:
```diff
@@ -73,7 +73,7 @@
         "symfony/uid": "^6.4|^7.0",
         "symfony/web-link": "^6.4|^7.0",
-        "symfony/webhook": "^7.2",
+        "symfony/webhook": "^7.3",
```
Are you sure?

lyrixx replied:
This was a typo. Thanks.

lyrixx commented:
PR updated, the failure is not related

fabpot commented:
Thank you @lyrixx.
